cloud-engineering-with-gcp/python/content/lab-03/README.md

```diff
@@ -0,0 +1,320 @@
+# Build and Deploy a Container on Google Cloud Run
+
+In our last lab, we deployed some static HTML files to a Google Cloud Storage bucket. In the third lab of this workshop, we're going create a Docker image to run our website on NGINX and deploy it to run on [Google Cloud Run](https://cloud.google.com/run/).
+
+## Step 1 &mdash; Create a Docker Image
+
+We'll use the [Pulumi Docker provider](https://www.pulumi.com/registry/packages/docker/) to build a Docker image we can push to Google Container Registry (GCR).
+
+Before we add anything to our project, let's make sure we configure the Docker CLI to push to GCR. We can do that by running the following command:
+
+```bash
+gcloud auth configure-docker
+```
+
+We'll also need to install the [Pulumi Command provider](https://www.pulumi.com/registry/packages/command/) so we can build our image. We can do this using `pip`. Add the following line to the bottom of your `requirements.txt`:
+
+```text
+pulumi_command>=0.0.0,<1.0.0
+```
+
+Install your dependencies using pip:
+
+```bash
+pip install -r requirements.txt
+```
+
+> If you'd like to do all of this from the command line in a single command, run the following:
+>
+> ```bash
+> echo "pulumi_command>=0.0.0,<1.0.0 >> requirements.txt && pip install -r requirements.txt`
+> ```
+
+Next, we need to import the Command provider's SDK. Add the following to the top of your `__main__.py` along with your other imports:
+
+```python
+import pulumi_command as command
+```
+
+Add the following to your `__main.py__` to build the Docker image:
+
+```python
+image_tag = pulumi.Output.concat(
```

Comment:
Are you creating the container registry or artifact registry in the code somewhere? You need a registry to push the code to. And how's the IAM set up on that registry? I'd define that in the code if this is a 200-level workshop.